File: DSA/Python/Questions/Graph/PathDecisions/CountPath.py
"""
# Question: Count All Possible Paths
# Link: https://leetcode.com/problems/number-of-ways-to-arrive-at-destination/

# Time Complexity: O(E * logV)
# Space Complexity: O(V + E)

# Algorithm:
# 1. Modified Dijkstra's algorithm
# 2. Track count of shortest paths
# 3. Handle modulo arithmetic
"""

from collections import defaultdict
import heapq


class CountPath:
    def count_paths(self, n: int, roads: list[list[int]]) -> int:
        MOD = 10**9 + 7
        graph = defaultdict(list)

        # Build adjacency list
        for u, v, time in roads:
            graph[u].append((v, time))
            graph[v].append((u, time))

        # Distance and ways arrays
        dist = [float("inf")] * n
        ways = [0] * n
        dist[0] = 0
        ways[0] = 1

        # Priority queue: (distance, node)
        pq = [(0, 0)]

        while pq:
            d, u = heapq.heappop(pq)

            if d > dist[u]:
                continue

            for v, time in graph[u]:
                if dist[v] > d + time:
                    dist[v] = d + time
                    ways[v] = ways[u]
                    heapq.heappush(pq, (dist[v], v))
                elif dist[v] == d + time:
                    ways[v] = (ways[v] + ways[u]) % MOD

        return ways[n - 1]


def main():
    solution = CountPath()
    roads = [[0, 1, 1], [0, 2, 2], [1, 2, 1]]
    print(solution.count_paths(3, roads))


if __name__ == "__main__":
    main()

File: DSA/src/Questions/Graph/CountPath.java
package Questions.Graph;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.PriorityQueue;

public class CountPath {
    static int countPaths(int n, ArrayList<ArrayList<Integer>> roads) {

        // Creating an adjacency list for the given graph.
        ArrayList<ArrayList<Pair>> adj = new ArrayList<>();
        for (int i = 0; i < n; i++) {
            adj.add(new ArrayList<>());
        }
        int m = roads.size();
        for (int i = 0; i < m; i++) {
            adj.get(roads.get(i).get(0)).add(new Pair(roads.get(i).get(1), roads.get(i).get(2)));
            adj.get(roads.get(i).get(1)).add(new Pair(roads.get(i).get(0), roads.get(i).get(2)));
        }

        // Defining a priority queue (min heap).
        PriorityQueue<Pair> pq = new PriorityQueue<>((x, y) -> x.first - y.first);

        // Initializing the dist array and the ways array
        // along with their first indices.
        int[] dist = new int[n];
        int[] ways = new int[n];
        Arrays.fill(dist, Integer.MAX_VALUE);
        Arrays.fill(ways, 0);
        dist[0] = 0;
        ways[0] = 1;
        pq.add(new Pair(0, 0));

        // Define modulo value
        int mod = (int) (1e9 + 7);

        // Iterate through the graph with the help of priority queue
        // just as we do in Dijkstra's Algorithm.
        while (!pq.isEmpty()) {
            int dis = pq.peek().first;
            int node = pq.peek().second;
            pq.remove();

            for (Pair it : adj.get(node)) {
                int adjNode = it.first;
                int edW = it.second;

                // This ‘if’ condition signifies that this is the first
                // time we’re coming with this short distance, so we push
                // in PQ and keep the no. of ways the same.
                if (dis + edW < dist[adjNode]) {
                    dist[adjNode] = dis + edW;
                    pq.add(new Pair(dis + edW, adjNode));
                    ways[adjNode] = ways[node];
                } else if (dis + edW == dist[adjNode]) {
                    // If we again encounter a node with the same short distance
                    // as before, we simply increment the no. of ways.
                    ways[adjNode] = (ways[adjNode] + ways[node]) % mod;
                }
            }
        }
        // Finally, we return the no. of ways to reach
        // (n-1)th node modulo 10^9+7.
        return ways[n - 1] % mod;
    }
    
    // Time Complexity: O( E* log(V)) { As we are using simple Dijkstra’s algorithm
    // here, the time complexity will be or the order E*log(V)}
    // Space complexity: O(N)

    public static void main(String[] args) {

        int n = 7;
        ArrayList<ArrayList<Integer>> roads = new ArrayList<ArrayList<Integer>>() {
            {
                add(new ArrayList<Integer>(Arrays.asList(0, 6, 7)));
                add(new ArrayList<Integer>(Arrays.asList(0, 1, 2)));
                add(new ArrayList<Integer>(Arrays.asList(1, 2, 3)));
                add(new ArrayList<Integer>(Arrays.asList(1, 3, 3)));
                add(new ArrayList<Integer>(Arrays.asList(6, 3, 3)));
                add(new ArrayList<Integer>(Arrays.asList(3, 5, 1)));
                add(new ArrayList<Integer>(Arrays.asList(6, 5, 1)));
                add(new ArrayList<Integer>(Arrays.asList(2, 5, 1)));
                add(new ArrayList<Integer>(Arrays.asList(0, 4, 5)));
                add(new ArrayList<Integer>(Arrays.asList(4, 6, 2)));

            }
        };

        System.out.println(countPaths(n, roads));
    }

}
